### PR TITLE
fix(unique): collision-free tuple key shared by intake and merge, loud on un-keyable types

### PR DESCRIPTION
**Pull request by ragnorc**
## Summary

Follow-up hardening on top of #133. That PR made composite `@unique(a, b)` a true tuple and introduced a shared `loader::composite_unique_key(parts)` that joins per-column scalars with `U+001F`, routing **both** the intake path (`enforce_unique_constraints_intra_batch`) and the branch-merge path (`exec/merge.rs::update_unique_constraints`) through it. That already closed the original `|` vs `U+001F` separator drift on `main`.

This PR takes the shared keying the rest of the way to correct-by-design — three improvements over the merged state:

1. **Collision-free by construction.** The shared key was a `U+001F`-joined string; a value containing that control char could still collide. The key is now the *tuple* of per-column scalar strings (`Vec<String>`), keyed directly. No separator means no data value can forge a collision, for any data.

2. **One scalar converter across both paths.** Intake rendered scalars with an explicit type-match; merge rendered them with Arrow's `array_value_to_string`. Both now derive the key through the same `composite_unique_key(group_columns, row)`, so the two paths can't drift on conversion either.

3. **Loud on un-keyable types.** The scalar converter returned `None` for any Arrow type it didn't recognize, and the caller treated `None` as null-exempt — so a `@unique` on a column type it couldn't reduce (a list, a blob) was silently un-enforced. It now returns `Err`, surfacing the constraint it can't enforce instead of weakening it in silence (deny-list: no silent failures).

The loop and error shapes stay per-path (intake fails fast within a batch; merge collects every conflict across the candidate stream). Only the key derivation is shared — the part that must never drift. Single-field `@unique`/`@key` behavior and error messages are unchanged.

## Tests

- `composite_unique_key_is_consistent_across_intake_and_merge` (`tests/consistency.rs`): loads two rows unique on `@unique(a, b)` where a column carries a literal `|`, on a branch, then merges. Pins that intake and merge key the tuple identically.
- loader unit tests: `composite_unique_key_builds_tuple_and_exempts_null` (tuple keying, null exemption, and `("x|y","z")` vs `("x","y|z")` distinctness) and `unique_key_scalar_errors_loudly_on_unkeyable_type` (a binary column now errors instead of silently passing).
- Full `cargo test --workspace --locked` green.

## Scope

Intra-batch and merge-candidate-set uniqueness, same as before. Cross-version uniqueness against already-committed rows on direct writes stays a documented known gap (`docs/dev/invariants.md`), unchanged by this PR.

---

> [!NOTE]
> **Medium Risk**
> Changes how uniqueness is computed on load and branch-merge write paths; incorrect keying could accept duplicates or reject valid rows, though behavior is pinned by new tests and scope stays intra-batch/merge-candidate checks only.
> 
> **Overview**
> Hardens composite `@unique` / `@key` enforcement by replacing separator-joined string keys with a **`Vec<String>` tuple** in dedup maps, so values containing `|`, `U+001F`, or other delimiters cannot collide.
> 
> **Intake** (`enforce_unique_constraints_intra_batch`) and **branch merge** (`update_unique_constraints`) now both call the same **`loader::composite_unique_key(group_columns, row)`**, which builds the tuple via **`unique_key_scalar`** instead of merge using `array_value_to_string` separately.
> 
> **`unique_key_scalar`** returns **`Err`** for non-null, un-keyable Arrow types (e.g. binary/blob) instead of treating unknown types like null and skipping enforcement; it also handles **Utf8, LargeUtf8, and Utf8View** string encodings.
> 
> Adds unit tests for tuple distinctness, loud failures on un-keyable types, and string encodings; adds **`composite_unique_key_is_consistent_across_intake_and_merge`** in `consistency.rs`. Updates **`docs/dev/invariants.md`** to describe the shared tuple keying.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3b910530bcaae96f3759cc6f9b8f3e289b19e114. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Greptile Summary</h3>

This PR hardens composite `@unique`/`@key` enforcement by replacing the U+001F-separator string key with a direct `Vec<String>` tuple, unifying key derivation into a single `loader::composite_unique_key(group_columns, row)` shared by both the intake and branch-merge paths, and turning silent `None` for un-keyable Arrow types into a loud `Err`.

- **Collision-free tuple key**: `HashMap<String, …>` → `HashMap<Vec<String>, …>` in both `enforce_unique_constraints_intra_batch` and `update_unique_constraints`; no separator means no data value can forge a key collision.
- **Shared scalar converter**: `unique_key_scalar` replaces the old `scalar_to_string` (intake) and `array_value_to_string` (merge), adding `LargeStringArray` and `StringViewArray` arms so Lance's physical string type can't cause a drift regression.
- **Loud on un-keyable types**: a `@unique` on a blob or list column now returns `Err` at write time instead of silently exempting every row from the check; three new unit tests and one cross-path integration test pin all three improvements.

<h3>Confidence Score: 5/5</h3>

Safe to merge — constraint enforcement is strictly tighter and the two previously-divergent code paths now share a single key derivation function verified by a round-trip integration test.

Both changed write paths (intake and merge) produce identical keys under the new implementation, the integration test confirms no phantom conflict on the pipe-character edge case, and all three string encodings Lance can emit are now handled.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| crates/omnigraph/src/loader/mod.rs | Core refactor: `composite_unique_key` now takes `(group_columns, row)` directly and returns `Result<Option<Vec<String>>>` (tuple); `unique_key_scalar` gains `LargeStringArray`/`StringViewArray` arms and turns unknown types into `Err` rather than silent `None`; group columns are resolved once outside the row loop. |
| crates/omnigraph/src/exec/merge.rs | `update_unique_constraints` adopts the shared `composite_unique_key`; `seen` maps change from `HashMap<String, String>` to `HashMap<Vec<String>, String>`; `row_signature` (divergent-update detection) is unchanged and correctly continues to use U+001F join. |
| crates/omnigraph/tests/consistency.rs | Adds `composite_unique_key_is_consistent_across_intake_and_merge`: loads two rows with literal `|` in composite key values, verifies intake and merge both accept them without a phantom conflict. |
| crates/omnigraph/src/table_store.rs | One-line comment update adjusting a stale line-number reference — no logic change. |
| docs/dev/invariants.md | Unique constraints row updated to document the shared tuple derivation and loud-on-un-keyable behaviour; cross-version gap note preserved. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Write / Load batch] --> B[enforce_unique_constraints_intra_batch]
    C[branch_merge candidate scan] --> D[update_unique_constraints]
    B --> E["Resolve group_columns once per constraint"]
    D --> F["Resolve group_columns once per constraint per batch"]
    E --> G["composite_unique_key(group_columns, row)"]
    F --> G
    G --> H{any column null?}
    H -- Ok/None --> I[Row exempt]
    H -- Ok/Some --> J["Insert Vec<String> key into dedup map"]
    H -- Err --> K[Fail loudly: unsupported type]
    J --> L{collision?}
    L -- No --> M[Continue]
    L -- Yes --> N["Intake: Err / Merge: MergeConflict"]
```

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Acrates%2Fomnigraph%2Fsrc%2Floader%2Fmod.rs%3A1468-1477%0AThe%20%60key.clone%28%29%60%20here%20happens%20on%20every%20row%2C%20including%20the%20vast%20majority%20that%20are%20not%20duplicates.%20The%20clone%20is%20only%20needed%20in%20the%20error%20branch%20%28to%20keep%20%60key%60%20alive%20for%20%60format_tuple%28%26key%29%60%20after%20the%20move%20into%20the%20map%29.%20Using%20the%20%60Entry%60%20API%20eliminates%20the%20allocation%20in%20the%20common%20path%3A%20%60entry.key%28%29%60%20makes%20the%20stored%20key%20available%20for%20the%20error%20message%20without%20a%20prior%20clone.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20%20%20%20%20match%20seen.entry%28key%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20std%3A%3Acollections%3A%3Ahash_map%3A%3AEntry%3A%3AOccupied%28e%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20let%20prev_row%20%3D%20*e.get%28%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20return%20Err%28OmniError%3A%3Amanifest%28format!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22%40unique%20violation%20on%20%7B%7D.%7B%7D%3A%20value%20'%7B%7D'%20appears%20in%20rows%20%7B%7D%20and%20%7B%7D%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type_name%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20format_tuple%28columns%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20format_tuple%28e.key%28%29%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20prev_row%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20row%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%29%29%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20std%3A%3Acollections%3A%3Ahash_map%3A%3AEntry%3A%3AVacant%28e%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20e.insert%28row%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%60%60%60%0A%0A&repo=modernrelay%2Fomnigraph&pr=160&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a>

<sub>Reviews (2): Last reviewed commit: ["fix(unique): cover all string encodings;..."](https://github.com/modernrelay/omnigraph/commit/3b910530bcaae96f3759cc6f9b8f3e289b19e114) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36480480)</sub>

### CODE REVIEW

File: crates/omnigraph/src/loader/mod.rs [hunk 1]
```diff
@@ -1482,66 +1480,94 @@ pub(crate) fn enforce_unique_constraints_intra_batch(
...
 }
 
-/// Render a unique constraint's columns for error messages: a single column
-/// as `col`, a composite as `(a, b)`.
-fn format_unique_columns(columns: &[String]) -> String {
-    match columns {
+/// Render a constraint's column tuple for error messages: a single item as
+/// `col`, a composite as `(a, b)`. Used for both the column list and the
+/// offending value tuple, which share the same shape.
+pub(crate) fn format_tuple(items: &[String]) -> String {
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/loader/mod.rs]: Done in 3b91053 — made it a private `fn`. It's only used inside `loader/mod.rs`, matching the visibility of the `format_unique_columns` it replaced.

File: crates/omnigraph/src/loader/mod.rs [hunk 2]
```diff
@@ -1482,66 +1480,94 @@ pub(crate) fn enforce_unique_constraints_intra_batch(
...
     if let Some(a) = array.as_any().downcast_ref::<Date64Array>() {
-        return Some(a.value(row).to_string());
+        return Ok(Some(a.value(row).to_string()));
     }
-    None
+    Err(OmniError::manifest(format!(
+        "uniqueness key: unsupported column type {:?} for @unique/@key enforcement",
+        array.data_type()
+    )))
 }
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/loader/mod.rs:1582]: Valid and important — this is the sharp edge of the loud-on-unknown-type change. Fixed in 3b91053: added `LargeStringArray` (LargeUtf8) and `StringViewArray` (Utf8View) arms, so a legal string column is keyable in every physical Arrow encoding. The `Err` path now fires only for a genuinely un-keyable logical type (list/blob/vector), never for a legal value that merely arrived in an unenumerated encoding. Pinned by a new unit test (`unique_key_scalar_handles_all_string_encodings`).

Today the engine pins String→Utf8 and the round-trip reads back as `StringArray` (the consistency test confirms), so this is forward-looking robustness against a future Lance/Arrow physical-type change — but it removes the false-positive class you identified.
